### PR TITLE
Added the lift subsystem 

### PR DESCRIPTION
**Pull request by Mboesn**

### CODE REVIEW

**Comment by HilelAmiel** [comment]
First of all change the pull request to somthing noraml.

File: src/main/java/frc/robot/SubSystems/Lift.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+import edu.wpi.first.wpilibj.Encoder;
+import edu.wpi.first.wpilibj.SpeedControllerGroup;
+import edu.wpi.first.wpilibj.command.Subsystem;
+import frc.robot.RobotComponents;
+/** This is the subsystem of the lift */
+public class Lift extends Subsystem {
+  private SpeedControllerGroup motors = RobotComponents.Lift.LIFT_MOTORS;
+  private Encoder encoder = RobotComponents.Lift.LIFT_ENCODER;
+
+public void setMotors(double speed){
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/SubSystems/Lift.java]: setMotorsPower()

File: src/main/java/frc/robot/SubSystems/Lift.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+import edu.wpi.first.wpilibj.Encoder;
+import edu.wpi.first.wpilibj.SpeedControllerGroup;
+import edu.wpi.first.wpilibj.command.Subsystem;
+import frc.robot.RobotComponents;
+/** This is the subsystem of the lift */
+public class Lift extends Subsystem {
+  private SpeedControllerGroup motors = RobotComponents.Lift.LIFT_MOTORS;
+  private Encoder encoder = RobotComponents.Lift.LIFT_ENCODER;
+
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/SubSystems/Lift.java:45]: Constructor pls? init motors and encoder there.

File: src/main/java/frc/robot/SubSystems/Lift.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+public void setMotors(double speed){
+  motors.set(speed);
+}
+public double getEncoder(){
+  return this.encoder.getDistance();
+}
+public void setEncoderDPP(double dpp){
+  this.encoder.setDistancePerPulse(dpp);
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/SubSystems/Lift.java]: remove and add this.encoder.setDistancePerPulse(dpp); to constractor

File: src/main/java/frc/robot/SubSystems/Lift.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+}
+  @Override
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/SubSystems/Lift.java:46]: Reset Encoder()

File: src/main/java/frc/robot/SubSystems/Lift.java (hunk 2 above)
Comment (HilelAmiel) [inline src/main/java/frc/robot/SubSystems/Lift.java:45]: Add two limit switches and all the changes requeired

File: src/main/java/frc/robot/RobotMap.java [hunk 5]
```diff
@@ -2,9 +2,11 @@
 
 /** This class defines the ports of the robot */
 public class RobotMap {
-    
+    //TODO: add real values
     // Device ID's for devices connected to the CAN bus
     static class CAN {
+        public static final int LIFT_MOTER1 = 0;
+        public static final int LIFT_MOTOR2 = 1;
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/RobotMap.java]: REAR, FRONT

File: src/main/java/frc/robot/RobotMap.java [hunk 6]
```diff
@@ -2,9 +2,11 @@
 
 /** This class defines the ports of the robot */
 public class RobotMap {
-    
+    //TODO: add real values
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/RobotMap.java]: nice :) 👍

File: src/main/java/frc/robot/RobotComponents.java [hunk 7]
```diff
@@ -1,13 +1,26 @@
...
 
-    public static class Drivetrain {
-        public static final TalonSRX REAR_RIGHT_MOTOR = new TalonSRX(RobotMap.CAN.REAR_RIGHT_MOTOR);
-        etc.
+    /*
+     * Add your class components here, and remove comment. example:
+     * 
+     * public static class Drivetrain { public static final WPI_TalonSRX
+     * REAR_RIGHT_MOTOR = new TalonSRX(RobotMap.CAN.REAR_RIGHT_MOTOR); etc. }
+     */
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/RobotComponents.java]: remove comment...

File: src/main/java/frc/robot/RobotComponents.java [hunk 8]
```diff
@@ -1,13 +1,26 @@
...
+    public static class Lift {
+        public static final WPI_TalonSRX LIFT_MOTOR1 = new WPI_TalonSRX(RobotMap.CAN.LIFT_MOTER1);
+        public static final WPI_TalonSRX LIFT_MOTOR2 = new WPI_TalonSRX(RobotMap.CAN.LIFT_MOTOR2);
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/RobotComponents.java]: REAR_LIFT_MOTOR
FRONT_LIFT_MOTOR

**Review by HilelAmiel — CHANGES_REQUESTED** [review]
You forgot to add your SS to Robot class also!

File: src/main/java/frc/robot/SubSystems/Lift.java (hunk 1 above)
Comment (Mboesn) [reply src/main/java/frc/robot/SubSystems/Lift.java]: yep

File: src/main/java/frc/robot/SubSystems/Lift.java (hunk 2 above)
Comment (Mboesn) [reply src/main/java/frc/robot/SubSystems/Lift.java:45]: sure

File: src/main/java/frc/robot/SubSystems/Lift.java (hunk 3 above)
Comment (Mboesn) [reply src/main/java/frc/robot/SubSystems/Lift.java]: Did it

File: src/main/java/frc/robot/RobotConstants.java [hunk 9]
```diff
@@ -2,16 +2,18 @@
...
     /** All the robot pid values for monitoring */
     public static class RobotPIDSettings {
     }
-    
+
+    // TODO: set real values
     /** Constants for sensors on the robot */
     public static class Sensors {
+        public static final double LIFT_ENCODER_DPP = 0;
     }
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/RobotConstants.java:21]: // Dpp

File: src/main/java/frc/robot/RobotMap.java [hunk 10]
```diff
@@ -2,9 +2,11 @@
 
 /** This class defines the ports of the robot */
 public class RobotMap {
-    
+    // TODO: add real values
     // Device ID's for devices connected to the CAN bus
     static class CAN {
+        public static final int LIFT_MOTER_FRONT = 0;
+        public static final int LIFT_MOTOR_REAR = 1;
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/RobotMap.java:9]: MOTOR

File: src/main/java/frc/robot/SubSystems/Lift.java [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+import edu.wpi.first.wpilibj.command.Subsystem;
+import frc.robot.RobotComponents;
+import frc.robot.RobotConstants;
+
+/** This is the subsystem of the lift */
+public class Lift extends Subsystem {
+  private SpeedControllerGroup motors;
+  private Encoder encoder;
+  private DigitalInput switchTop, switchBottom;
+
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/SubSystems/Lift.java:15]: topSwitch...

File: src/main/java/frc/robot/SubSystems/Lift.java [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+    this.switchBottom = RobotComponents.Lift.LIFT_SWITCH_BOTTOM;
+    this.encoder.setDistancePerPulse(RobotConstants.Sensors.LIFT_ENCODER_DPP);
+  }
+
+  public void setMotorsPower(double power) {
+    this.motors.set(power);
+  }
+
+  public double getEncoder() {
+    return this.encoder.getDistance();
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/SubSystems/Lift.java]: getHeight()

File: src/main/java/frc/robot/SubSystems/Lift.java [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+  private SpeedControllerGroup motors;
+  private Encoder encoder;
+  private DigitalInput switchTop, switchBottom;
+
+  public Lift() {
+    this.motors = RobotComponents.Lift.LIFT_MOTORS;
+    this.encoder = RobotComponents.Lift.LIFT_ENCODER;
+    this.switchTop = RobotComponents.Lift.LIFT_SWITCH_TOP;
+    this.switchBottom = RobotComponents.Lift.LIFT_SWITCH_BOTTOM;
+    this.encoder.setDistancePerPulse(RobotConstants.Sensors.LIFT_ENCODER_DPP);
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/SubSystems/Lift.java]: this.encoder next to this.encoder

**Review by HilelAmiel — CHANGES_REQUESTED** [review]
Robot Class

**Review by HilelAmiel — CHANGES_REQUESTED** [review]
Encoders accroding to https://phoenix-documentation.readthedocs.io/en/latest/ch14_MCSensor.html

File: src/main/java/frc/robot/SubSystems/Lift.java [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+import edu.wpi.first.wpilibj.command.Subsystem;
+import frc.robot.RobotComponents;
+import frc.robot.RobotConstants;
+
+/** This is the subsystem of the lift */
+public class Lift extends Subsystem {
+  private SpeedControllerGroup motors;
+  private DigitalInput topSwitch, bottomSwitch;
+  private WPI_TalonSRX encoder;
+
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/SubSystems/Lift.java:15]: remove SpeedControllerGroup. change to TalonSRX

File: src/main/java/frc/robot/SubSystems/Lift.java [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+    return this.bottomSwitch.get();
+  }
+
+  public int getHeight() {
+    return this.encoder.getSelectedSensorPosition() / RobotConstants.Sensors.LIFT_ENCODER_TPD
+        + RobotConstants.Sensors.LIFT_ENCODER_OFFSET;
+  }
+
+  public void restEncoderHeight() {
+    this.encoder.setSelectedSensorPosition(0);
```
Comment (HilelAmiel) [inline src/main/java/frc/robot/SubSystems/Lift.java]: resetEncoder()

**Review by HilelAmiel — CHANGES_REQUESTED** [review]
good job last things

**Review by HilelAmiel — APPROVED** [review]
Done!

**Pull request merged**